### PR TITLE
Hcx communication

### PR DESCRIPTION
**Pull request by khavinshankar**
Added HCX communication flow

Backend PR: https://github.com/coronasafe/care/pull/1346

_Further docs will be added later in docs repo_

### CODE REVIEW

File: src/Components/HCX/ClaimCard.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  useLayoutEffect(() => {
+    if (cardContainerRef.current) {
+      setContainerDimensions({
+        width: cardContainerRef.current.offsetWidth,
+        height: cardContainerRef.current.offsetHeight,
+      });
+    }
+  }, [cardContainerRef]);
+
+  console.log(containerDimensions);
```
Comment (rithviknishad) [inline src/Components/HCX/ClaimCard.tsx]:
```suggestion
```

File: src/Components/HCX/ClaimCardCommunication.tsx [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+  index?: number;
+}
+
+export default function ClaimCardCommunication({
+  claim,
+  setShowMessages,
+}: IProps) {
+  const dispatch = useDispatch<any>();
+  const [messages, setMessages] = useState<IMessage[]>([]);
+  const [responses, setResponses] = useState<IMessage[]>([]);
```
Comment (rithviknishad) [inline src/Components/HCX/ClaimCardCommunication.tsx]: TODO: replace useDispatch with useQuery/request

File: src/Components/HCX/SendCommunicationModal.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  show: boolean;
+  onClose: () => void;
+}
+
+export default function SendCommunicationModal({
+  communication,
+  ...props
+}: Props) {
+  const dispatch = useDispatch<any>();
+  const [isLoading, setIsLoading] = useState(false);
```
Comment (rithviknishad) [inline src/Components/HCX/SendCommunicationModal.tsx]: TODO: replace useDispatch with useQuery/request

File: src/Redux/actions.tsx [hunk 4]
```diff
@@ -168,6 +172,31 @@ export const HCXActions = {
...
+    partialUpdate(id: string, obj: Partial<HCXCommunicationModel>) {
+      return fireRequest("partialUpdateHCXCommunication", [], obj, {
+        external_id: id,
+      });
+    },
+    delete(id: string) {
+      return fireRequest("deleteHCXCommunication", [], {}, { external_id: id });
+    },
+  },
+
```
Comment (rithviknishad) [inline src/Redux/actions.tsx]: TODO: not required after migrating to useQuery/request.

```suggestion
```

File: src/Redux/actions.tsx [hunk 5]
```diff
@@ -187,4 +216,8 @@ export const HCXActions = {
   makeClaim(claim: string) {
     return fireRequest("hcxMakeClaim", [], { claim });
   },
+
+  sendCommunication(communication: string) {
+    return fireRequest("hcxSendCommunication", [], { communication });
+  },
```
Comment (rithviknishad) [inline src/Redux/actions.tsx]: TODO: not required after migrating to useQuery/request

```suggestion
```

File: src/Redux/api.tsx [hunk 6]
```diff
@@ -1655,6 +1658,41 @@ const routes = {
...
+
+  deleteHCXCommunication: {
+    path: "/api/v1/hcx/communication/{external_id}/",
+    method: "DELETE",
+  },
+
+  hcxSendCommunication: {
+    path: "/api/v1/hcx/send_communication/",
+    method: "POST",
+  },
```
Comment (rithviknishad) [inline src/Redux/api.tsx]: TODO: `TRes` and `TBody` to be defined for it to work with useQuery/request

File: src/Redux/api.tsx [hunk 7]
```diff
@@ -1655,6 +1663,57 @@ const routes = {
...
+  getHCXCommunication: {
+    path: "/api/v1/hcx/communication/{external_id}/",
+    method: "GET",
+    TRes: Type<HCXCommunicationModel>(),
+  },
+
+  updateHCXCommunication: {
+    path: "/api/v1/hcx/communication/{external_id}/",
+    method: "PUT",
+    TRes: Type<HCXCommunicationModel>(),
```
Comment (rithviknishad) [inline src/Redux/api.tsx:1695]: Missing TBody

File: src/Redux/api.tsx [hunk 8]
```diff
@@ -1655,6 +1663,57 @@ const routes = {
...
+  },
+
+  partialUpdateHCXCommunication: {
+    path: "/api/v1/hcx/communication/{external_id}/",
+    method: "PATCH",
+    TRes: Type<HCXCommunicationModel>(),
```
Comment (rithviknishad) [inline src/Redux/api.tsx:1701]: Missing TBody

**Review by rithviknishad — CHANGES_REQUESTED** [review]

**Review by rithviknishad — APPROVED** [review]

**Comment by nihal467** [comment]
@khavinshankar can you update the PR description for review purpose to understand what changed in this PR

**Comment by nihal467** [comment]
image

- [x] unable to send the message because the payor-app is down, HCX is not live

image
image

- [x] After we type a message, we need to click the add button and then the submit button to send the message. This is kind of confusing because I believe this feature will be used to seek additional documents and should have an expected behavior of chat, single-click to send a message, and an icon on the text box to select documents for attachment purposes to the chat. 

image
image

- [ ] Compared to the existing staging, the two fields insurer ID and name are replaced with one single field named insurer in the HCX, so we should be keeping a consistent type of data in the platform, so how to handle this?

**Comment by nihal467** [comment]
@gigincg @khavinshankar update me, for re-reviewing once the things are sorted

**Comment by khavinshankar** [comment]
New UI
image

**Comment by nihal467** [comment]
image

- [ ] the entire page is not responsive

image

- [ ] rename the cancel button as close, since we can add multiple insurance in a single flow, so the pop-up wont close automatically after adding one insurance

image

- [ ] add a warning if notification is disabled, as realtime updation only happens if the notifications are enabled

image

- [ ] height of the column should be dynamically changed to occupy the option selected from the dropdown

**Comment by nihal467** [comment]
image
image
image

@khavinshankar Even though the pre-auth was successful from the care side, the swasth payor app is not showing the patient's name in the preauth section, but the patient name is visible in the coverage eligibility list

**Comment by khavinshankar** [comment]
As the issues mentioned above involve working on the entire HCX module, it will be tackled in a separate PR and will be monitored in a new issue #8413. This way HCX Communication flow is not blocked for merging.

_Ideally the above-mentioned issues will be tackled along with HCX plug seperation._

**Comment by nihal467** [comment]
@khavinshankar can you get this PR  ready for review and testing

**Comment by bodhish** [comment]
@khavinshankar is this required in the context of NHCX?

**Comment by khavinshankar** [comment]
@bodhish Yes, this is one of the HCX flow but this PR can be closed as this is already part of #8420

**Comment by bodhish** [comment]
Let's close it 🚀

**Comment by gigincg** [comment]
Closing based on confirmation from @khavinshankar

**Pull request closed without merging**